### PR TITLE
Blocked cyclic based multi-GPU blocked cholesky + clone_here bug fix

### PR DESCRIPTION
**Pull request by nicelhc13**
This PR contains two items:
1) blocked cyclic based multi-GPU blocked cholesky using row partitioning
2) `clone_here` fix; `cupy.as_array()` could not handle multiple CUDA stream well. This PR replace it with asynchronous copy specifying a target stream. `_GPUMemory` managed CUDA device object and multiple threads shared this global object without synchronization. This PR removed on-demand CUDA context change while `clone_here()` with an assumption that each task is only assigned a single GPU. Note that these are not the final fixes. We expect `cupy` will open their fixes soon and we will move the context object to thread local memory to avoid data race.

### CODE REVIEW

File: parla/cuda.py
Comment (insertinterestingnamehere) [inline parla/cuda.py:55]: I think you mean for this to be asarray_async? sync refers to the cuda semantics that we're trying to avoid.

File: parla/cuda.py
Comment (insertinterestingnamehere) [inline parla/cuda.py:66]: Just pushed an alternate fix for the context manager issue to the other PR. LMK what you think. Otherwise, LGTM.

File: parla/cuda.py
Comment (nicelhc13) [reply parla/cuda.py:55]: Yes my bad. I put synchronize() before right after the copy. That's why I put that name. I am going to fix this name.

File: parla/cuda.py
Comment (nicelhc13) [reply parla/cuda.py:66]: Fixed!

File: parla/cuda.py
Comment (nicelhc13) [reply parla/cuda.py]: Fixed

File: examples/blocked_cholesky_gpu.py
Comment (arthurp) [inline examples/blocked_cholesky_gpu.py:67]: I think the style guides for Python generally have a space after `#` in comments, so there is no reason to remove all said spaces. ;-) This is not important though. Just a nit.

File: examples/blocked_cholesky_multigpu.py
Comment (arthurp) [inline examples/blocked_cholesky_multigpu.py:190]: These functions can be defined in a class along with `cholesky_blocked_inplace` so the class can capture the value of `num_gpu` and clean up this code a bit.

Also I think the mess here means that a multi-device array abstraction type would be useful. This simple approach may well not be general enough, but we clearly need a way to work with and directly index arrays that are actually multiple arrays on different devices. User code should never have to have junk like this.

File: examples/blocked_cholesky_multigpu.py
Comment (arthurp) [inline examples/blocked_cholesky_multigpu.py:243]: Why not use `@spawn` and get some parallelism here. :-)

File: parla/array.py
Comment (arthurp) [inline parla/array.py:8]: We should avoid this if at all possible. IIRC, importing cupy here prevents running the code at all without CUDA installed.

File: parla/cuda.py
Comment (arthurp) [reply parla/cuda.py:66]: The code in this PR is actually wrong because this function can be called with a different device active to copy data to a remote device instead of to the current device. Let's go with Ian's solution if at all possible.

File: parla/cuda.py
Comment (arthurp) [inline parla/cuda.py:67]: Use `isinstance` for type checks in Python.

File: parla/array.py
Comment (arthurp) [inline parla/array.py:134]: This isn't important since we are in a hurry, but please clean your diffs to remove random changes like this. In this case it actually makes the code match the style guides less well.

**Review by arthurp — APPROVED** [review]
I left some important and some less important comments. But overall, LGTM.

File: parla/cuda.py
Comment (arthurp) [reply parla/cuda.py:66]: I see in https://github.com/ut-parla/Parla.py/pull/20#issuecomment-810648174 that Ian's fix is not enough. So do whatever hack you need but make sure to document that this is breaking the semantics of the function in a big way, so we can go back and fix it.

File: examples/blocked_cholesky_gpu.py
Comment (insertinterestingnamehere) [reply examples/blocked_cholesky_gpu.py:67]: This is correct. The spaces should be there.

File: examples/blocked_cholesky_multigpu.py
Comment (insertinterestingnamehere) [reply examples/blocked_cholesky_multigpu.py:190]: Yah, we need to think about how to do this better. A multi-device array abstraction is a huge code cost, so we might be better served by working out some kind of hook into dask to leverage what they've already implemented in some way. Alternatively we could work on expanding our partitioning functionality. IDK. It's worth thinking about.

**Comment by insertinterestingnamehere** [comment]
I just double-checked, and I think all the remaining comments here are non-blocking. Should I just go ahead and merge this?

**Comment by arthurp** [comment]
> I just double-checked, and I think all the remaining comments here are non-blocking. Should I just go ahead and merge this?

Can you fix the following quickly before merging:

* https://github.com/ut-parla/Parla.py/pull/21#discussion_r604498388
* https://github.com/ut-parla/Parla.py/pull/21#discussion_r604499617
* https://github.com/ut-parla/Parla.py/pull/21#discussion_r604501673 (Adding TODOs to track the semantic problem)

I don't want to lose those issues because they can bite us later.

**Comment by insertinterestingnamehere** [comment]
@nicelhc13 do you mind taking care of those?

**Comment by nicelhc13** [comment]
Sure guys. I cannot work on this now since the filesystem of the TACC servers are not working now; I have no multi-GPU servers except them now so I cannot test it. I will merge this ASAP when the servers work.

File: examples/blocked_cholesky_multigpu.py
Comment (nicelhc13) [reply examples/blocked_cholesky_multigpu.py:243]: It looks very good point. But for now, if you don't mind, let me just postpone this work to the another PR because it is excluded from time measurement and some people are waiting for this PR.

**Comment by insertinterestingnamehere** [comment]
LGTM now.

**Comment by insertinterestingnamehere** [comment]
Thank you!

**Pull request merged**
